### PR TITLE
Try to fix macOS hang when building Python 2

### PR DESCRIPTION
**Pull request by BillyONeal**
* Don't open multiple handles to the same file when looking for absolute paths.
* Don't use `std::future`.

### CODE REVIEW

**Comment by BillyONeal** [comment]
I kicked off 3 copies of the osx-x64 build to see if this helps:

* https://dev.azure.com/vcpkg/public/_build/results?buildId=99494&view=results
* https://dev.azure.com/vcpkg/public/_build/results?buildId=99495&view=results
* https://dev.azure.com/vcpkg/public/_build/results?buildId=99495&view=results

**Comment by BillyONeal** [comment]
> I kicked off 3 copies of the osx-x64 build to see if this helps:
> 
> * https://dev.azure.com/vcpkg/public/_build/results?buildId=99494&view=results
> * https://dev.azure.com/vcpkg/public/_build/results?buildId=99495&view=results
> * https://dev.azure.com/vcpkg/public/_build/results?buildId=99495&view=results

Initial results are promising since all 3 of these got past Python2 without issues but without a full understanding of the cause of the problem it is difficult to be sure.

File: include/vcpkg/base/parallel-algorithms.h [hunk 1]
```diff
@@ -1,85 +1,163 @@
...
     {
-        const size_t thread_count = static_cast<size_t>(get_concurrency());
-        const size_t num_threads = std::max(static_cast<size_t>(1), std::min(thread_count, work_count));
+        F work;
+        size_t work_count;
+        std::atomic<size_t> next_offset;
 
-        std::vector<std::future<void>> workers;
-        workers.reserve(num_threads - 1);
+        WorkCallbackContext(F init_f, size_t work_count) : work(init_f), work_count(work_count), next_offset(0) { }
```
Comment (ras0219-msft) [inline include/vcpkg/base/parallel-algorithms.h:26]:
```suggestion
        WorkCallbackContext(F init_f, size_t work_count) : work(std::move(init_f)), work_count(work_count), next_offset(0) { }
```

File: include/vcpkg/base/parallel-algorithms.h [hunk 2]
```diff
@@ -1,85 +1,163 @@
...
 
-        for (size_t i = 0; i < num_threads - 1; ++i)
+        void run()
         {
-            workers.emplace_back(std::async(std::launch::async | std::launch::deferred, [&work]() { work(); }));
+            auto offset = next_offset.load(std::memory_order_relaxed);
+            while (offset < work_count)
+            {
+                if (!next_offset.compare_exchange_weak(offset, offset + 1, std::memory_order_relaxed))
```
Comment (ras0219-msft) [inline include/vcpkg/base/parallel-algorithms.h:33]: Could this use `fetch_add` instead, since we can assume that `work_count + threads < SIZE_MAX`?

File: include/vcpkg/base/parallel-algorithms.h [hunk 3]
```diff
@@ -1,85 +1,163 @@
...
-            while (i < c.size())
+        WorkCallbackContext<F> context{work, work_count};
+        PtpWork ptp_work([](PTP_CALLBACK_INSTANCE,
+                            void* context,
+                            PTP_WORK) noexcept { static_cast<WorkCallbackContext<F>*>(context)->run(); },
+                         &context,
+                         nullptr);
+        if (ptp_work)
+        {
+            auto max_threads = (std::min)(work_count, static_cast<size_t>(get_concurrency()));
```
Comment (ras0219-msft) [inline include/vcpkg/base/parallel-algorithms.h:90]: Since this is leveraging the thread pool, should this be larger than the hardware concurrency and letting the thread pool decide the appropriate oversubscription based on CPU vs IO?

File: include/vcpkg/base/parallel-algorithms.h [hunk 4]
```diff
@@ -1,85 +1,163 @@
...
         }
 
-        std::atomic_size_t next{0};
-
-        execute_in_parallel(c.size(), [&]() {
-            size_t i = 0;
-            while (i < c.size())
+        WorkCallbackContext<F> context{work, work_count};
+        const size_t bg_threads = std::min(work_count, static_cast<size_t>(get_concurrency())) - 1;
+        std::vector<Optional<JThread>> workers(bg_threads);
```
Comment (ras0219-msft) [inline include/vcpkg/base/parallel-algorithms.h:133]: Why use Optional/resize/emplace vs reserve/emplace_back? Exception safety?

File: src/vcpkg/base/files.cpp [hunk 5]
```diff
@@ -1453,25 +1455,64 @@ namespace vcpkg
...
+            }
+
+            output.append(buffer, this_read);
+            if (!this->eof())
+            {
+                Checks::unreachable(VCPKG_LINE_INFO, "Got a partial read without an error or end");
+            }
         }
 
+        ec.clear();
```
Comment (ras0219-msft) [inline src/vcpkg/base/files.cpp:1515]: Why is this needed? I thought that `!ec` was a precondition for APIs accepting error code out parameters.

File: src/vcpkg/base/files.cpp [hunk 6]
```diff
@@ -1453,25 +1455,64 @@ namespace vcpkg
...
+                if (this_read != buffer_size)
+                {
+                    break;
+                }
+
+                output.append(buffer, this_read);
+            }
+
+            auto maybe_error = ::ferror(m_fs);
+            if (maybe_error)
```
Comment (ras0219-msft) [inline src/vcpkg/base/files.cpp:1501]:
```suggestion
            if (auto maybe_error = this->error_raw())
```

File: src/vcpkg/base/files.cpp [hunk 7]
```diff
@@ -1453,25 +1455,64 @@ namespace vcpkg
...
+                output.clear();
+                return output;
+            }
+
+            output.append(buffer, this_read);
+            if (!this->eof())
+            {
+                Checks::unreachable(VCPKG_LINE_INFO, "Got a partial read without an error or end");
+            }
         }
```
Comment (ras0219-msft) [inline src/vcpkg/base/files.cpp:1513]: Could this common chunk of "append the rest of the stream without transformation" be deduplicated between `read_to_end` and `best_effort_read_contents_if_shebang`?

**Review by ras0219-msft — APPROVED** [review]

File: include/vcpkg/base/parallel-algorithms.h (hunk 4 above)
Comment (BillyONeal) [reply include/vcpkg/base/parallel-algorithms.h:133]: emplace_back requires that T be movable, thread is not.

File: src/vcpkg/base/files.cpp (hunk 5 above)
Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:1515]: Nope, !ec is a postcondition for APIs accepting error_code out parameters. There are no preconditions.

File: include/vcpkg/base/parallel-algorithms.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/base/parallel-algorithms.h:26]: This is a functor/callable, which are usually pointers, and for which std::move is a pessimization. (This is also why they are passed by value)

File: include/vcpkg/base/parallel-algorithms.h (hunk 2 above)
Comment (BillyONeal) [reply include/vcpkg/base/parallel-algorithms.h:33]: I'm not sure we *can* make that assumption. (`fetch_add` was @Thomas1664 's initial solution and I requested this form to avoid overflow)

I suppose we can assume this is unlikely and add other edge cases for it though I'm not sure that would be better... stand by

File: include/vcpkg/base/parallel-algorithms.h (hunk 3 above)
Comment (BillyONeal) [reply include/vcpkg/base/parallel-algorithms.h:90]: get_concurrency is `VCPKG_MAX_CONCURRENCY`, not hardware_concurrency. I'm inclined to do what the user said here unless we have concrete data saying we should try more.

File: include/vcpkg/base/parallel-algorithms.h (hunk 2 above)
Comment (Thomas1664) [reply include/vcpkg/base/parallel-algorithms.h]: > Could this use `fetch_add` instead, since we can assume that `work_count + threads < SIZE_MAX`?

`work_count` could reach `SIZE_MAX`. I remember that some PRs that used smaller types in context of the install plan were rejected.

File: include/vcpkg/base/parallel-algorithms.h (hunk 4 above)
Comment (Thomas1664) [reply include/vcpkg/base/parallel-algorithms.h]: > emplace_back requires that T be movable, thread is not.

`std::thread` is movable: https://en.cppreference.com/w/cpp/thread/thread/thread

File: include/vcpkg/base/parallel-algorithms.h (hunk 1 above)
Comment (ras0219-msft) [reply include/vcpkg/base/parallel-algorithms.h:28]: I understand why taking it by rvalue reference would be a pessimization, but how would std::move reduce performance?

File: include/vcpkg/base/parallel-algorithms.h [hunk 8]
```diff
@@ -1,85 +1,168 @@
...
+        WorkCallbackContext<F> context{work, work_count};
+        PtpWork ptp_work([](PTP_CALLBACK_INSTANCE,
+                            void* context,
+                            PTP_WORK) noexcept { static_cast<WorkCallbackContext<F>*>(context)->run(); },
+                         &context,
+                         nullptr);
+        if (ptp_work)
+        {
+            auto max_threads = (std::min)(work_count, static_cast<size_t>(get_concurrency()));
+            max_threads = (std::min)(max_threads, SIZE_MAX - work_count); // to avoid overflow in fetch_add
```
Comment (ras0219-msft) [inline include/vcpkg/base/parallel-algorithms.h:94]:
```suggestion
            if (SIZE_MAX - work_count > max_threads) Checks::unreachable(VCPKG_LINE_INFO);
```
I don't think trying to be fancy here to support more than 18,446,744,073,709,550,000-ish (or 4,294,966,000-ish) items is worth it.

Also, I think the current line is pessimistic by 1 item/thread -- if `work_count == SIZE_MAX`, 1 thread is permissible.

**Review by ras0219-msft — APPROVED** [review]

File: include/vcpkg/base/parallel-algorithms.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/base/parallel-algorithms.h:28]: `std::move` forces that the value be bound to an rvalue-reference. This is why it, for example, breaks NRVO in `return std::move(blah);` cases. It is likely that it wouldn't matter in this case; I'm just used to 'pass the functors by value' resulting from the mess that caused us to add `_Pass_fn`: passing functors by reference broke the ability to inline through function pointers. Maybe my paranoia here is out of date by now but given that we control all the functors here and aren't going to ever pass ones that are expensive to copy I'm inclined to keep following the convention.

https://github.com/microsoft/STL/blob/bd3d740ae5de7255c720b8133c5d23aa131e0760/stl/inc/xutility#L558-L584

File: include/vcpkg/base/parallel-algorithms.h (hunk 8 above)
Comment (BillyONeal) [reply include/vcpkg/base/parallel-algorithms.h:94]: >I don't think trying to be fancy here to support more than 18,446,744,073,709,550,000-ish (or 4,294,966,000-ish) items is worth it.

I think it is worth it, at least as long as 32 bit is not dead. (And arm-linux is still very much a thing, even if we won't get much parallelism there)

>Also, I think the current line is pessimistic by 1 item/thread -- if work_count == SIZE_MAX, 1 thread is permissible.

Sure. (Note that `work_count == SIZE_MAX` is impossible because the `end()` element can't exist in the address space. :) )

Comment (BillyONeal) [reply include/vcpkg/base/parallel-algorithms.h]: (Moreover, a dependency from this stuff on Checks:: makes this more complex, not less)
